### ISSUE TITLE
net

### ISSUE THREAD

**Issue by congmonkey** [bob48523/MultiTask-Learning#2]
"""
Created on Sat Aug 12 15:59:58 2017

@author: chenjiaxu
"""

import torch
import torch.nn as nn
import torch.nn.functional as F
import math

class SE(nn.Module):
    def __init__(self, in_planes):
        super(SE, self).__init__()
        self.fc1 = nn.Linear(in_planes, in_planes//16)
        self.relu = nn.ReLU(inplace=True)
        self.fc2 = nn.Linear(in_planes//16, in_planes)
        self.sigmoid = nn.Sigmoid()
        
    def forward(self, x):
        #global FLOPS
        w = F.avg_pool2d(x, kernel_size=x.size(2))
        w = self.fc1(w.view(w.size(0),-1))
        w = self.relu(w)
        #FLOPS += self.fc1.in_features*self.fc1.out_features
        w = self.fc2(w)
        w = self.sigmoid(w)
        #FLOPS += self.fc2.in_features*self.fc2.out_features
        w = w.view(w.size(0),w.size(1),1,1)
        w = w.repeat(1,1,x.size(2),x.size(3))
        out = w*x      
        return out

class PreActBlock(nn.Module):
    '''Pre-activation version of the original Bottleneck module.'''
    widen_factor = 1
    
    def __init__(self, in_planes, planes, stride=1):
        super(PreActBlock, self).__init__()
        
        
        self.bn1 = nn.BatchNorm2d(in_planes)
        self.relu = nn.ReLU(inplace=True)
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        
        self.bn2 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes*self.widen_factor, kernel_size=3, padding=1, bias=False)
        
        self.bn3 = nn.BatchNorm2d(planes*self.widen_factor)
        self.se = SE(planes*self.widen_factor)

        if stride != 1 or in_planes != planes*self.widen_factor:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, planes*self.widen_factor, kernel_size=1, stride=stride, bias=False)             
            )
            

    def forward(self, x):
        #global FLOPS
        out = self.bn1(x)
        #out = self.relu(out)
        shortcut = self.shortcut(out) if hasattr(self, 'shortcut') else x
        
        out = self.conv1(out)     
        #FLOPS += self.conv1.in_channels*self.conv1.out_channels*self.conv1.kernel_size[0]*self.conv1.kernel_size[1]*out.size(2)*out.size(3)

        out = self.bn2(out) 
        out = self.relu(out)
        out = self.conv2(out)
        out = self.bn3(out)
        #FLOPS += self.conv2.in_channels*self.conv2.out_channels*self.conv2.kernel_size[0]*self.conv2.kernel_size[1]*out.size(2)*out.size(3)                
       
        out = self.se(out)
              
        out += shortcut
        return out

class PreActBottleneck(nn.Module):
    '''Pre-activation version of the original Bottleneck module.'''
    widen_factor = 4
    
    def __init__(self, in_planes, planes, stride=1):
        super(PreActBottleneck, self).__init__()
        

        self.bn1 = nn.BatchNorm2d(in_planes)
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=1, bias=False)
        
        self.bn2 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        
        self.bn3 = nn.BatchNorm2d(planes)
        self.conv3 = nn.Conv2d(planes, planes*self.widen_factor, kernel_size=1, bias=False)   
        
        self.bn4 = nn.BatchNorm2d(planes*self.widen_factor)
        self.se = SE(planes*self.widen_factor)

        if stride != 1 or in_planes != planes*self.widen_factor:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, planes*self.widen_factor, kernel_size=1, stride=stride, bias=False)             
            )
            
            

    def forward(self, x):
        #global FLOPS
        out = self.bn1(x)
        shortcut = self.shortcut(out) if hasattr(self, 'shortcut') else x
        
        out = self.conv1(out) 
        #FLOPS += self.conv1.in_channels*self.conv1.out_channels*self.conv1.kernel_size[0]*self.conv1.kernel_size[1]*out.size(2)*out.size(3)
        
        out = self.conv2(F.relu(self.bn2(out)))
        #FLOPS += self.conv2.in_channels*self.conv2.out_channels*self.conv2.kernel_size[0]*self.conv2.kernel_size[1]*out.size(2)*out.size(3)        
    
        out = self.conv3(F.relu(self.bn3(out)))
        #FLOPS += self.conv3.in_channels*self.conv3.out_channels*self.conv3.kernel_size[0]*self.conv3.kernel_size[1]*out.size(2)*out.size(3)
        out = self.bn4(out)
        
        out = self.se(out)
              
        out += shortcut
        return out

class Resnet_sen_share(nn.Module):
    def __init__(self, block, num_blocks, cifar_classes, tiny_classes, vgg_classes):
        super(Resnet_sen_share, self).__init__()
        self.in_planes = 32
        width = [32, 64, 128, 256]
        
        self.base = nn.Sequential(
            nn.Conv2d(3, self.in_planes, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(self.in_planes),
            nn.ReLU(inplace=True),
            self._make_layer(block, width[0], num_blocks[0], stride=1),
            self._make_layer(block, width[1], num_blocks[1], stride=2),
            self._make_layer(block, width[2], num_blocks[2], stride=2),
            self._make_layer(block, width[3], num_blocks[3], stride=2),
            nn.AvgPool2d(8),
        )
#        self.conv1 = nn.Conv2d(3, self.in_planes, kernel_size=3, padding=1, bias=False)
#        self.bn1 = nn.BatchNorm2d(self.in_planes)
#        self.relu = nn.ReLU(inplace=True)
        
         
#        self.layer1 = self._make_layer(block, width[0], num_blocks[0], stride=1)
#        self.layer2 = self._make_layer(block, width[1], num_blocks[1], stride=2)
#        self.layer3 = self._make_layer(block, width[2], num_blocks[2], stride=2)
#        self.layer4 = self._make_layer(block, width[3], num_blocks[3], stride=2)
    
        self.fc_cifar = nn.Linear(width[3]*block.widen_factor, cifar_classes)
        self.fc_tiny = nn.Linear(width[3]*block.widen_factor, tiny_classes)
        self.fc_vgg = nn.Linear(width[3]*block.widen_factor, vgg_classes)
        
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
                m.weight.data.normal_(0, math.sqrt(2. / n))
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1)
                m.bias.data.zero_()
            elif isinstance(m, nn.Linear):
                m.bias.data.zero_()

    def _make_layer(self, block, planes, num_blocks, stride):
        strides = [stride] + [1]*(num_blocks-1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_planes, planes, stride))
            self.in_planes = planes * block.widen_factor
        return nn.Sequential(*layers)

    def forward(self, x_cifar, x_tiny, x_vgg):
        #global FLOPS
        #FLOPS = 0
#        out = F.relu(self.bn1(self.conv1(x)))
#        #FLOPS += self.conv1.in_channels*self.conv1.out_channels*self.conv1.kernel_size[0]*self.conv1.kernel_size[1]*out.size(2)*out.size(3)
#        
#        out = self.layer1(out)
#        out = self.layer2(out)
#        out = self.layer3(out)
#        out = self.layer4(out)       
#        out = F.avg_pool2d(out, 8)

        out_cifar = self.base(x_cifar)
        out_tiny = self.base(x_tiny)
        out_vgg = self.base(x_vgg)
        
        out_cifar = out_cifar.view(out_cifar.size(0), -1)
        out_tiny = out_tiny.view(out_tiny.size(0), -1)
        out_vgg = out_vgg.view(out_vgg.size(0), -1)
        
        out_cifar = self.fc_cifar(out_cifar)
        out_tiny = self.fc_tiny(out_tiny)
        out_vgg = self.fc_vgg(out_vgg)
        #FLOPS += self.fc.in_features*self.fc.out_features
        return out_cifar, out_tiny, out_vgg
